tests/discovery.rs

```diff
@@ -153,6 +153,47 @@ macro_rules! generate_tests {
             assert_eq!(client.get("/"), "hello");
             assert_eq!(client.get("/bye"), "bye");
         }
+
+        #[test]
+        fn outbound_error_reconnects_after_backoff() {
```

Comment:
Thanks for adding a test for this (and I appreciate the amount of comments in the test, as well).

However, I wonder if we might also want to have a more granular unit test for just the backoff behaviour as well? Since the failure to notify the backoff future is masked by the way the various control futures are `join`ed and are thus all polled, the backoff future still gets polled when it's inside of `Proxy::run`, maybe we would also want to test the backoff future in isolation.

I'm imagining something along the lines of:

```rust

struct MockService {
    polls: usize,
    succeed_after: usize,
}

impl MockService {
    fn new(succeed_after: usize) -> Self {
          Self { polls: 0, succeed_after, }
    }
}

impl Service for MockService {
    type Request = ();
    type Response = ();
    type Error = ();
    type Future = FutureResult<(), ()>;

    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
        self.polls += 1;
        if self.polls > self.succeed_after {
            Ok(Async::Ready(()))
        } else {
            Err(())
        }
    }

    fn call(&mut self, req: Self::Request) -> Self::Future {
        if self.polls > self.succeed_after {
            future::ok(())
        } else {
            future::err(())
        }
    }
}

#[test]
fn test_backoff() {
    let mock = MockService::new(2);
    let mut backoff = Backoff::new(mock, Duration::from_millis(5)); // or whatever.
    let mut rt = tokio::runtime::current_thread::Runtime::new()
        .unwrap();
    let call = rt.block_on_for(backoff.call(), Duration::from_millis(15));
    assert!(call.is_ok());
    assert_eq!(mock.polls, 3);
}

```